examples/empirical_sinkhorn_div/script.jl

```diff
@@ -0,0 +1,103 @@
+# [GPC18]: Aude Genevay, Gabriel Peyré, Marco Cuturi, Learning Generative Models with Sinkhorn Divergences, Proceedings of the Twenty-First International Conference on Artficial Intelligence and Statistics, (AISTATS) 21, 2018
+# [FSV+19]: Feydy, Jean, et al. "Interpolating between optimal transport and MMD using Sinkhorn divergences." The 22nd International Conference on Artificial Intelligence and Statistics. PMLR, 2019.
+#
+# Like the Sinkhorn loss, the Sinkhorn divergence is smooth and convex in both of its arguments. However, the Sinkhorn divergence is unbiased -- i.e. $S_{\varepsilon}(\mu, \nu) = 0$ iff $\mu = \nu$. 
+#
+# Unlike previous examples, here we demonstrate a learning problem similar to Figure 1 of Feydy et al. over *empirical measures*, i.e. measures that have the form $\mu = \frac{1}{N} \sum_{i = 1}^{N} \delta_{x_i}$ where $\delta_x$ is the Dirac delta function at $x$. 
+# 
+# We first load packages. 
+#
+using OptimalTransport
+using ReverseDiff
+using Distributions
+using LinearAlgebra
+using Distances
+using Plots
+using Logging
+using Optim
+
+# As a ground truth distribution, we set $\rho$ to be a Gaussian mixture model with `k = 5` components, equally spaced around a circle, and sample an empirical distribution of size $N$, $\mu \sim \rho$.
+
+k = 3
+d = 2
+θ = π * range(0, 2(1 - 1 / k); length=k)
+μ = 2 * hcat(sin.(θ), cos.(θ))
+ρ = MixtureModel(MvNormal[MvNormal(x, 0.25 * I) for x in eachrow(μ)])
+N = 100
+μ_spt = rand(ρ, N)'
+scatter(μ_spt[:, 1], μ_spt[:, 2]; markeralpha=0.25, title=raw"$\mu$")
+
+# Now, suppose we want to approximate $\mu$ with another empirical distribution $\nu$, i.e. we want to minimise $\nu \mapsto \operatorname{S}_{\varepsilon}(\mu, \nu)$ over possible empirical distributions $\nu$. In this case we have $M$ particles in $\nu$, which we initialise following a Gaussian distribution. 
+M = 100
+ν_spt = rand(M, d);
+# Assign uniform weights to the Diracs in each empirical distribution. 
+μ = fill(1 / N, N)
+ν = fill(1 / M, M);
+
+# Since $\mu$ is fixed, we pre-compute the cost matrix $C_{\mu}$. 
+C_μ = pairwise(SqEuclidean(), μ_spt');
+
+# Define the loss function to minimise, where `x` specifies the locations of the Diracs in $\nu$. 
+# 
+# We will be using `ReverseDiff` with a precompiled tape. For this reason, we need the Sinkhorn algorithm to perform a fixed number of (e.g. 50) iterations.
+# Currently, this can be achieved by setting `maxiter = 50` and `atol = rtol = 0` in calls to `sinkhorn` and `sinkhorn_divergence`. 
+function loss(x, ε)
+    C_μν = pairwise(SqEuclidean(), μ_spt', x')
+    C_ν = pairwise(SqEuclidean(), x')
+    return sinkhorn_divergence(
+        μ, ν, C_μν, C_μ, C_ν, ε; maxiter=50, atol=rtol = 0, regularization=true
+    )
+end
+# Set entropy regularisation parameter
+ε = 1.0;
+
+# Use ReverseDiff with a precompiled tape and Optim.jl to minimise $\nu \mapsto \operatorname{S}_{\varepsilon}(\mu, \nu)$. Note that this is problem is *not* convex, so we find a local minimium. 
```

Comment:
It's not convex because we are optimising over the positions of the points, not the weights. E.g. we can move from one local minima to another by exchanging points.

Reply:
True, I missed that these points are fixed in the notation of Theorem 1 in the paper.